google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -398,6 +423,14 @@ public void nack() {
             @Override
             public void run() {
               try {
+                if (ackHandler.totalExpiration.isBefore(now())) {
+                  // Message expired while waiting. We don't extend these messages anymore,
+                  // so it was probably sent to someone else. Don't work on it.
+                  // Don't nack it either, because we'd be nacking someone else's message.
```

Comment:
Is this true? Does the server hand out the same ackId each time it hands out the same message?
Without thinking too hard about how PubSub is built, that surprises me. Why isn't ackId == messageId in that case?

Reply:
The second question is easier to answer: You can have multiple subscriptions to the same topic. Acking a message in one subscription has no effect on the other subscription. If they were the same, we can't tell what subscription you're acking from.

The spec is silent on this. Generally, I don't think you can rely on the ack ID changing but you also cannot rely on it being the same. At least as of a couple months ago, the ID stays the same when you pull message multiple times. I'm sorry for being intentionally vague because I'm not sure how much info I can share. If you'd like, I can reach out to pubsub to see what they can comment publicly.

Reply:
Interesting, we have support contract etc with GCP so I can follow it up with our contacts 👍